frameworks/FSharp/giraffe/src/App/StatefullRendering.fs

```diff
@@ -0,0 +1,60 @@
+﻿namespace App
+module rec StetefullRendering =
+
+    open System.Text
+    open Giraffe.GiraffeViewEngine
+    open System.Net
+    open System.IO
+
+    let inline private add (str:string) (target: StreamWriter) =
+        target.Write str
+        target
+
+    let inline private add' (str:string) (target: StreamWriter) =
+        target.Write str
+
+    let private closingBracket = ">"
+
+    let private writeStartElement target elemName (attributes : XmlAttribute array) : unit =
+
+        match attributes with
+        | [||] -> 
+            do target
+                |> add "<"
+                |> add elemName
+                |> add' closingBracket
+        | _    ->
+            do target |> add "<" |> add' elemName
+
+            for attr in attributes do
+                match attr with
+                | KeyValue (k, v) -> 
+                    target |> add " " |> add k |> add "=\"" |> add' (WebUtility.HtmlEncode v)
+                | Boolean k -> 
+                    target |> add " " |> add' k
+
+            do target |> add' closingBracket
+
+    let private writeEndElement target elemName = 
+        do target |> add "</" |> add elemName |> add' ">"
+
+    let private writeParentNode target ((name, attrs) : XmlElement) (nodes : XmlNode list) =
+        do writeStartElement target name attrs 
+        do nodes |> List.iter (writeHtmlNode target)
+        do name  |> writeEndElement target
+
+    let rec writeHtmlNode (target: StreamWriter) (node : XmlNode)  =
+        match node with
+        | EncodedText text -> do target |> add' (WebUtility.HtmlEncode text)
+        | RawText text -> do target |> add' text
+        | ParentNode (e, nodes) -> do writeParentNode target e nodes
+        | VoidElement (n, attrs) -> do writeStartElement target n attrs
+    
+    let renderHtmlToStream (encoding:Encoding) node = 

```

Comment:
Why is there a separate memory stream being created rather then using the asp.net core buffers in response? The Memory stream is allocating a resizing byte array internally so writing into it, only to later flush into the response is just waste. we should look to write directly into the response body stream, example of a similar implementation to above is here [ByteViewEngine](https://github.com/gerardtoconnor/ViewEngineTests/blob/master/ByteViewEngine.fs) and outperforms most due to writing directly body stream. Is there a important reason MS is being used I have overlooked?

Reply:
Writing to the http out would mean I need to use task api for every write, which assuming recursive walk would result in more allocations for tasks and binds than memory stream where I could write without the need of async in the first place.

The other bit is that I am not entirely sure about the consequences of writing to the http out stream without knowing the content length upfront, I think kestrell would fallback into chunked content encoding, @benaadams might give a clue here.

Reply:
This is why I provided the example of my ByteViewEngine, it does not use async api and therefore only makes small compromise on blocking. The writing of the small chunks will be hotpath into the kestrel buffers with no blocking until the buffer fills and needs to flush, the buffer is big enough and pages small enough that there will be minimal buffer flushes, on the flushes, unlike a DB roundtrip or file read request, there is no waiting on the return so minimal blocking. The whole operation can be a task, but it is performant on a single sync thread. 

On the content length, not sure of the upfront content length omission, that is a little trickier but doable if needed. 

We can pool & reuse the memory streams, .SetLength(0) will reset it once done, a more complex concise but painful way ... If we pre-compute & Cache the template length with no variables at start, we can iterate the variables to calculate their additional size each time before writing out to the buffer. the double iteration wont add much more time as the array (List) will already be CPU cached from prior run and was just loaded anyway from DB call, importantly no extra big MS array allocs.

Either way, upfront content length setting is not simple so if anyone has insight into the chunked content encoding, would save us wasting time on any complex work arounds.

Reply:
Thanks @benaadams , this is helpful, fully agree, more often then not, DB calls/driver are the bottleneck and can make these other micro optimisations immaterial.

I think we should try avoid the chunking in principal, (despite being thin layer) it's bit lazy/cheat when pooling a MemoryStream isn't too difficult. I prefer the pooled MS over SB given there is wasted CPU & memory on converting to UTF-16 string, when the ultimate output is UTF-8, best to write & store as UTF-8 as the c# middleware test inefficiently is having to iterate & convert the SB string (utf-16) twice to first count the bytes (UTF-8), and then write the string (utf-16) to UTF-8.

The idea around pre-computing the base template size and then at runtime only calculating the incremental GetByteCount Sum of variables is probably a level of runtime complexity that will be slower then simply pooling buffers so can be ignored.

@dv00d00 do you think MemoryStream pooling way to go then, means you can leave current templating engine as is, just getting/returning MS from pool? It's probably reasonable to pre-allocate pool on start up (50 MemoryStreams of size 1024 bytes?).

Reply:
> fully agree, more often then not, DB calls/driver are the bottleneck and can make these other micro optimisations immaterial.

aspnetcore still can output ~3.12Gb/s; which is 30% of network capacity; h20 is at ~4.46Gb/s or 45%; assuming the Db transfer is similar they are both 60% - 90% of the network. Possibly for the next round that's bumping up to 100Gb/s network, so that may all change.

Also aspnetcore platform -> aspnetcore mvc on the same db driver is a 64% drop in perf; so its probably worth tweaking that last little bit out 😉

Reply:
> Also aspnetcore platform -> aspnetcore mvc on the same db driver is a 64% drop in perf; so its probably worth tweaking that last little bit out 😉

That's what we did with Giraffe, MVC being the last little bit out 😉. 

With Giraffe we will continue to micro optimise don't worry, in addition to utilizing all the biggest wins in aspnetcore, it's best of both worlds for us, you're doing all the heavy lifting!

Reply:
Results for current (merged) version of giraffe are available https://twitter.com/ben_a_adams/status/1008406784212656130